### PR TITLE
feat(framework) Add `get_run_status` and `update_run_status` to `LinkState`

### PR DESCRIPTION
**Pull request by panh99**
- Amended `LinkState`.
- Implemented new methods for `InMemoryLinkState` and `SqliteLinkState`.
- Added unit test for the new methods.
- Updated the codebase accordingly.

### CODE REVIEW

File: src/py/flwr/server/superlink/state/in_memory_state.py [hunk 1]
```diff
@@ -17,18 +17,35 @@
...
+
+@dataclass
+class RunRecord:
+    """The record of a specific run, including its status and timestamps."""
+
+    run: Run
+    status: RunStatus
+    starting_at: str
+    running_at: str
+    finished_at: str
```
Comment (mohammadnaseri) [inline src/py/flwr/server/superlink/state/in_memory_state.py:48]: Are `starting_at` and `finished_at` are the times? shouldn't we use float for them?
What is `running_at`?

File: src/py/flwr/server/superlink/state/in_memory_state.py [hunk 2]
```diff
@@ -338,7 +366,54 @@ def get_run(self, run_id: int) -> Optional[Run]:
...
+                    status.status,
+                    new_status.status,
+                )
+                return False
+
+            # Check if the sub-status is valid
+            if not has_valid_sub_status(status):
+                log(
+                    ERROR,
+                    'Invalid run status: "%s:%s"',
```
Comment (mohammadnaseri) [inline src/py/flwr/server/superlink/state/in_memory_state.py:403]: shouldn't we mention it as invalid run sub-status?

File: src/py/flwr/server/superlink/state/sqlite_state.py [hunk 3]
```diff
@@ -795,6 +793,75 @@ def get_run(self, run_id: int) -> Optional[Run]:
...
+        rows = self.query(query, (sint64_run_id,))
+
+        # Check if the run_id exists
+        if not rows:
+            log(ERROR, "`run_id` is invalid")
+            return False
+
+        # Check if the status transition is valid
+        row = rows[0]
+        status = RunStatus(
```
Comment (mohammadnaseri) [inline src/py/flwr/server/superlink/state/sqlite_state.py:827]: Should we change its name to `current_status` for better clarity?

File: src/py/flwr/common/constant.py [hunk 4]
```diff
@@ -120,3 +120,27 @@ class ErrorCode:
     def __new__(cls) -> ErrorCode:
         """Prevent instantiation."""
         raise TypeError(f"{cls.__name__} cannot be instantiated.")
+
+
+class Status:
+    """Run status."""
+
+    STARTING = "starting"
```
Comment (mohammadnaseri) [inline src/py/flwr/common/constant.py:128]: I find the `STARTING` status quite confusing. Does it mean that there is a duration between starting a run until running a run?
Also, is it a good idea that we use a status `RUNNING` for a run?

Comment (panh99) [reply src/py/flwr/common/constant.py:128]: Yes, it's actually described in the [RFC](https://www.notion.so/flowerlabs/Introduce-Run-status-3a131d67488c4226877d986a0e6a8550). It's the time when the superlink receives the `CreateRun` request from the SuperLink but the ServerApp process hasn't started yet.

File: src/py/flwr/server/superlink/state/in_memory_state.py (hunk 1 above)
Comment (panh99) [reply src/py/flwr/server/superlink/state/in_memory_state.py:48]: Yes, they are timestamps. The `running_at` is the time when the ServerApp starts.

File: src/py/flwr/server/superlink/state/sqlite_state.py (hunk 3 above)
Comment (panh99) [reply src/py/flwr/server/superlink/state/sqlite_state.py]: Sure, thanks.

File: src/py/flwr/server/superlink/state/in_memory_state.py (hunk 2 above)
Comment (panh99) [reply src/py/flwr/server/superlink/state/in_memory_state.py]: Yes, I've updated it.

File: src/py/flwr/server/superlink/linkstate/in_memory_linkstate.py [hunk 5]
```diff
@@ -401,7 +430,54 @@ def get_run(self, run_id: int) -> Optional[Run]:
             if run_id not in self.run_ids:
                 log(ERROR, "`run_id` is invalid")
                 return None
-            return self.run_ids[run_id]
+            return self.run_ids[run_id].run
+
+    def get_run_status(self, run_ids: set[int]) -> dict[int, RunStatus]:
```
Comment (jafermarq) [inline src/py/flwr/server/superlink/linkstate/in_memory_linkstate.py:435]: (very minor) Would it be better to assume `run_ids` is a list, and then construct a set out of it and proceed with the rest of the logic? The return dict will be the same but we'll prevent unnecessary iterations if `run_ids` contains duplicated ids for some reason.

File: src/py/flwr/server/superlink/linkstate/in_memory_linkstate.py [hunk 6]
```diff
@@ -26,13 +27,30 @@
...
+from .utils import (
+    generate_rand_int_from_bytes,
+    has_valid_sub_status,
+    is_valid_transition,
+    make_node_unavailable_taskres,
+)
+
+
+@dataclass
+class RunRecord:
```
[the remark this answers is not in the document]
Comment (panh99) [reply src/py/flwr/server/superlink/linkstate/in_memory_linkstate.py:46]: I am a bit unsure about this. For me, this `RunRecord` is nothing more than an ad-hoc solution for `InMemoryLinkState` storing run-related information. It's not a common type that will be used in multiple places. Wdyt? @jafermarq

File: src/py/flwr/common/constant.py [hunk 7]
```diff
@@ -128,3 +128,27 @@ class ErrorCode:
...
```
[the remark this answers is not in the document]
Comment (panh99) [reply src/py/flwr/common/constant.py:136]: Happy to discuss this during tmrw's meeting.

File: src/py/flwr/server/superlink/linkstate/in_memory_linkstate.py (hunk 5 above)
Comment (panh99) [reply src/py/flwr/server/superlink/linkstate/in_memory_linkstate.py:435]: That makes sense.

File: src/py/flwr/server/superlink/linkstate/in_memory_linkstate.py [hunk 8]
```diff
@@ -26,13 +27,30 @@
...
+    """The record of a specific run, including its status and timestamps."""
+
+    run: Run
+    status: RunStatus
+    starting_at: str
```
Comment (jafermarq) [inline src/py/flwr/server/superlink/linkstate/in_memory_linkstate.py:51]:
```suggestion
    pending_at: str
    starting_at: str
```

File: src/py/flwr/server/superlink/linkstate/in_memory_linkstate.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (jafermarq) [reply src/py/flwr/server/superlink/linkstate/in_memory_linkstate.py:46]: I agree with your logic!

File: src/py/flwr/server/superlink/fleet/vce/vce_api_test.py [hunk 9]
```diff
@@ -113,12 +115,22 @@ def register_messages_into_state(
...
+    state.run_ids[run_id] = RunRecord(
+        Run(
+            run_id=run_id,
+            fab_id="Mock/mock",
+            fab_version="v1.0.0",
+            fab_hash="hash",
+            override_config={},
+        ),
+        RunStatus(
+            status=Status.STARTING,
```
Comment (jafermarq) [inline src/py/flwr/server/superlink/fleet/vce/vce_api_test.py:127]:
```suggestion
            status=Status.PENDING,
```

Comment (jafermarq) [reply src/py/flwr/server/superlink/fleet/vce/vce_api_test.py:127]: better change it for consistency?

**Review by danieljanes — APPROVED** [review]

**Pull request merged**